### ISSUE TITLE
Option to limit runner to a particular run or job

### ISSUE THREAD

**Issue by simonbyrne** [actions/runner#620]
**Describe the enhancement**

In addition to supporting the `--once` option (#510), it would be useful if a runner could be limited to a specific run or job, exiting once this job is completed, or if cancelled or already completed. 

A typical use case would be for autoscaling: either via a webhook or polling the [workflow runs API](https://docs.github.com/en/rest/reference/actions#list-workflow-runs-for-a-repository), i would start up the necessary number of workers for the jobs/runs.

**Code Snippet**
```
./run.sh --job=<job_id> # run job matching job_id, or exit if complete/cancelled
```

**Additional information**
Similar to the [Buildkite](https://buildkite.com/docs/agent/v3/cli-start) `--acquire-job` option.

**Comment by jgoux** [comment]
It would be invaluable for us as well. Right now we have to maintain a "garbage collector" lambda which kills runners older than X hours if there are zombies runners.

**Comment by alexellis** [comment]
I didn't find this issue when searching, probably due to the language used in this issue.

Please see also: [Tie a specific workflow job to a specific ephemeral runner via labels](https://github.com/actions/runner/issues/2147) for a more detailed explanation of the problem this solves.

**Comment by arianvp** [comment]
Without this it's basically impossible to use `jit_config` without getting a lot of zombie instances when workflows get cancelled. I don't see how you can reliably use Just-in-time runners _without_ this feature. Please. this would be great to have!

**Comment by bduffany** [comment]
> Without this it's basically impossible to use jit_config without getting a lot of zombie instances when workflows get cancelled

I agree with this, and would like to mention the experience that I've had so far trying to work with the JIT runner feature.

My initial (simple) approach was:
* Listen for `workflow_job` `queued` events
* On each event, create a jit config and spawn a runner with the config.

In this approach, instead of matching up job IDs to runner IDs (which is currently not supported), we spawn N jobs for N runners and any runner can take any job. So the total count of `queued` events that we receive has should match the total count of runners that we spawn. But this is unreliable because:
* If a runner crashes, we are now permanently short one runner and will always be behind by 1 job.
* If a job is cancelled before a runner claims it, we will always have an extra runner that sits around for a very long time, consuming valuable compute resources.

I thought of a solution for these issues, but it is pretty complicated:
* Have a worker that periodically monitors the number of runners running vs the number of jobs available. If the number of runners is less than the number of available jobs, spawn a new runner.
  * This monitoring is not straightforward to implement; we need to do it in such a way to avoid consuming excessive API quota, while also being immediately responsive to webhook events.
* When starting the runner, also spawn a separate process which monitors the runner and does the following: if the runner does not claim a job after a certain (short) timeout, e.g. 5 minutes, then kill it.
  * This also seems like it will be awkward to implement - without forking the runner codebase, I think the best we can do is to scan the stdout/stderr of the runner and look for a message indicating that it has started a job. Also, there is still the downside that some compute is being consumed unnecessarily.

I wonder if the GitHub team had a different use case in mind when designing jit runners? The consequences of the current jit design has led me to a pretty complicated design in order to make use of it for my particular use case, and I wonder if I am missing something obvious in how jit runners are supposed to be used.

**Comment by alexellis** [comment]
I landed back here again. Does the GitHub runner team have an update? Would this be something that you would consider?

**Comment by hanwen-flow** [comment]
Is someone from github still reading this?

I am also interested in this feature request. In addition to the reasons mentioned by the previous posters, even within a set of uniform runners, there can be per-runner cache affinity: in build systems, the snapshot of post-submit successful build provides a great cache seed for an incremental pre-submit test run targeting the same branch. Exploiting this requires knowing something about the job (eg. the branch it is targeting) so we can setup the right snapshot before kicking off the agent.

**Comment by mikeclayton** [comment]
I've come here from the recent GitHub self-hosted runner pricing announcements at https://github.com/orgs/community/discussions/182186


Adding to https://github.com/actions/runner/issues/620#issuecomment-1956828993 above (@bduffany on Feb 21, 2024)

> * If the job does not exist by the time the runner starts (e.g. if it's cancelled) then the runner can immediately exit, resulting in minimal wasted compute resources.

I've seen duplicate webhooks invocations by the GitHub backend for the same workflow causing multiple runners to be spun up due to our custom orchestrator not de-duping job ids, so the runner's built-in check would ideally include "if the job does not exist **or is already running** then the runner can immediately exit,"...

**Comment by dubsalot** [comment]
Borrowing this from another comment:

> But in the case of GitHub runners, it is github.com which keeps track of jobs and which sends your runner jobs to do.


Maybe GitHub could provide a mechanism that allows customers to self-host their own queue as along with their runners.

**Comment by xheklinalepri022-design** [comment]
We are coming!!

**Comment by MrSuddenJoy** [comment]
+1 for this.

**Comment by shankiyani** [comment]
Given that JIT configs currently operate on the enterprise, org, and repo levels, is it possible to support a job level jit config? This could solve a few key issues. 

1. It further optimizes the job polling query and returns at most 1 result -- the job assigned to it. 
2. A runner could confidently kill itself if its assigned job is not present.
3. A job level jit config is scoped strictly to the secrets and permissions of a specific job.

The tradeoff is it cannot run alongside non-job level jit runners (so any of enterprise, org, or repo). Job dedicated runners will randomly find their jobs taken by the other runners. It shouldn't be a major issue in that the existing tiers also have this same issue if ran concurrently.

**Comment by aaronandkylie2025-hue** [comment]
./run.sh --job=<job_id> # run job matching job_id, or exit if complete/cancelled

**Comment by xheklinalepri022-design** [comment]
I dont not use. Yt hub


Il dom 1 mar 2026, 23:27 aaronandkylie2025-hue ***@***.***>
ha scritto:
